localstack/services/s3/codec.py

```diff
@@ -0,0 +1,119 @@
+import io
+from typing import IO, Any, Optional
+
+
+class AwsChunkedDecoder(io.RawIOBase):
+    """
+    This helper class takes a IO[bytes] stream, and decodes it on the fly, so that S3 can directly access the stream
+    without worrying about implementation details of `aws-chunked`.
+    It needs to expose the trailing headers, which will be available once the stream is fully read.
+    You can also directly pass the S3 Object, so the stream would set the checksum value once it's done.
+    TODO: should we register callback instead of passing the object?
+    See `aws-chunked` format here: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html
+    """
+
+    def readable(self):
+        return True
+
+    def __init__(
+        self, stream: IO[bytes], decoded_content_length: int, s3_object: Optional[Any] = None
+    ):
+        self._stream = stream
+
+        self._decoded_length = decoded_content_length  # Length of the encoded object
+        self._new_chunk = True
+        self._end_chunk = False
+        self._chunk_size = 0
+        self._trailing_headers = {}
+        self.s3_object = s3_object
+
+    @property
+    def trailing_headers(self):
+        return self._trailing_headers
```

Comment:
nit: You mentioned the callback above, but you could make it more explicit if you would raise an `ValueError` (or maybe there's something that's even more fitting) if the trailing headers are accessed even though they have not been read yet.
This would make it quite explicit and issues with missing trailing headers due to the stream not being fully read could be avoided.

Reply:
Agreed, that would be nice! I'm not directly accessing those trailing headers anymore, the previous commits were directly accessing them but it made a different code path between regular stream and aws-chunked one. But having this contract properly set would make things a bit more transparent I agree. I'm going to add this, and look for an exception to raise. Will change that 👍